### DISCUSSION TITLE
link exo platform community edition with an active directory

### DISCUSSION THREAD

**Discussion by xBlueDragonn** [exoplatform/exo-community#70] [Q&A]
Hi, i would like to link my exoplatform community edition install with docker-compose to an microsoft active directory server is this possible ?

**Comment by rdenarie** [comment]
Hello,
You should follow documentation about LDAP Integration : 
https://docs.exoplatform.org/administration/LDAP.html

Properties must be added in exo.properties file, and this file must be mounted in eXo container in /etc/exo/exo.properties

**Reply by xBlueDragonn** [reply]
So i have to create the directory exo in etc and create my exo.properties. After that in my docker-compose.yml i mount up exo.properties like that in the volumes section : - exo_properties:/etc/exo/exo.properties and also adding  exo_properties:
Is that the right way to do it ?

**Reply by rdenarie** [reply]
exo.properties must be a file on your local machine : for example /home/username/exo/conf/exo.properties.
Then, in docker-compose, in exo volumes part, you add a line like this : 
      - /home/username/conf/exo.properties:/etc/exo/exo.properties:ro
This will mount your local file exo.properties in exo container, in file /etc/exo/exo.properties

Then, in your local exo.properties file, you will be able to add ad properties to configure it, following the documentation

**Reply by xBlueDragonn** [reply]
it will create automally /etc/exo/exo.properties or i have to create it ? Because when i check in etc i have no directory named exo.

**Reply by rdenarie** [reply]
/etc/exo/exo.properties will be automatically created INSIDE the exo docker container.

> Because when i check in etc i have no directory named exo.
Where do you check ? on your locale machine ? or in the exo container ?

**Reply by xBlueDragonn** [reply]
so in my machine i will not see it, ok.

**Reply by xBlueDragonn** [reply]
when i check in the volume created by docker i do no find my file exo.properties and i also get the following error :État HTTP 500 – Erreur interne du serveur 
here is the log: 
2024-03-14 06:17:59,947 | ERROR | Servlet.service() for servlet [portal] in context with path [/portal] threw exception [o.a.c.c.C.[.[.[/portal].[portal]<http-nio-0.0.0.0-8080-exec-8>]
java.lang.NullPointerException: Cannot invoke "org.exoplatform.services.jcr.ext.app.SessionProviderService.getSessionProvider(Object)" because "providerService" is null
        at org.exoplatform.frameworks.jcr.web.ThreadLocalSessionProviderInitializedFilter.doFilter(ThreadLocalSessionProviderInitializedFilter.java:138)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:178)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:168)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:481)
        at org.gatein.sso.integration.SSODelegateValve.invoke(SSODelegateValve.java:142)
        at org.gatein.sso.agent.tomcat.ServletAccessValve.invoke(ServletAccessValve.java:55)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
        at org.gatein.sso.agent.tomcat.ServletAccessValve.invoke(ServletAccessValve.java:55)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:130)
        at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:765)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
        at org.apache.catalina.authenticator.SingleSignOn.invoke(SingleSignOn.java:241)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:390)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:926)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1790)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
        at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
        at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.base/java.lang.Thread.run(Thread.java:840)

**Reply by rdenarie** [reply]
Hello, Can you provide the full starting log ?
Thanks

**Reply by xBlueDragonn** [reply]
I fix the previous problem. I configure my exo.properties like the following screen :
![Capture d'écran 2024-03-18 123718](https://github.com/exoplatform/exo-community/assets/145013620/2efeae3e-7001-4996-908a-f518b727ebc4)
But when i connect to exoplatform and check the users i dont have any users of my domain.

**Reply by rdenarie** [reply]
can you provide the platform.log file ?

Can you also check that the exo.properties file is correctly mounted in your exo docker container ? 

Go inside the container, then check the exo.properties file, you should see your propertiesd

In addition, on my side, when I use an AD for some tests, I have to put this for admin.dn :
exo.ldap.admin.dn=ROOT\\Administrator, where ROOT is the AD domain.

In this case, you should see error logs in log file saying that credentials are not correct

**Reply by xBlueDragonn** [reply]
Here is the platform log 
[plat.log](https://github.com/exoplatform/exo-community/files/14633425/plat.log)

I also do a find -name exo.properties to find my file but when i cat it its empty 
![image](https://github.com/exoplatform/exo-community/assets/145013620/7a072b68-7901-461a-88da-251363d3781e)

**Reply by xBlueDragonn** [reply]
And this is how i mouted my exo.properties 
![image](https://github.com/exoplatform/exo-community/assets/145013620/65821b4f-cb29-478e-b1f1-53f74f3a0435)

**Reply by rdenarie** [reply]
To read files inside the container, you need to do 
docker exec -it containerName bash
Then you will have a bash inside the container, and you will be able to do
cat /etc/exo/exo.properties

This cat  should show the real content of your properties file

By reading logs It seems that your configuration file is not taken in account.

**Reply by xBlueDragonn** [reply]
here is the result 
![image](https://github.com/exoplatform/exo-community/assets/145013620/95823255-1b66-4fcf-a446-7c738879b6dd)
the exo.properties file have the config

**Reply by xBlueDragonn** [reply]
i got this error that pop up :
exo_1    | 2024-03-18 09:49:36,354 | INFO  | Indexing in bulk 4 documents [o.e.analytics.es.AnalyticsESClient<Analytics-ingestor-0>]
exo_1    | 2024-03-18 09:50:00,005 | INFO  | Start treating GroupSpaceBinding queue [o.e.s.c.b.job.QueueGroupSpaceBindingJob<ExoScheduler_Worker-17>]
exo_1    | 2024-03-18 09:50:00,011 | INFO  | End treating GroupSpaceBinding queue [o.e.s.c.b.job.QueueGroupSpaceBindingJob<ExoScheduler_Worker-17>]
exo_1    | 2024-03-18 09:50:00,050 | INFO  | Job portal:PORTAL.IDM.IDM.DATA.IMPORT threw a JobExecutionException:  [org.quartz.core.JobRunShell<ExoScheduler_Worker-16>]
exo_1    | org.quartz.JobExecutionException: An error occurred while executing job IDMEntitiesImportJob
exo_1    |      at org.exoplatform.services.organization.externalstore.job.IDMEntitiesImportJob.execute(IDMEntitiesImportJob.java:22)
exo_1    |      at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
exo_1    |      at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
exo_1    | Caused by: java.lang.RuntimeException: Operation error on LDAP store
exo_1    |      at org.exoplatform.services.organization.idm.externalstore.PicketLinkIDMExternalStoreService.lambda$new$0(PicketLinkIDMExternalStoreService.java:88)
exo_1    |      at org.exoplatform.services.organization.idm.externalstore.PicketLinkIDMExternalStoreService.executeOnExternalStore(PicketLinkIDMExternalStoreService.java:410)
exo_1    |      at org.exoplatform.services.organization.idm.externalstore.IDMExternalStoreListAccess.load(IDMExternalStoreListAccess.java:59)
exo_1    |      at org.exoplatform.services.organization.idm.externalstore.IDMExternalStoreListAccess.load(IDMExternalStoreListAccess.java:24)
exo_1    |      at org.exoplatform.services.organization.externalstore.IDMExternalStoreImportService.loadModifiedUsers(IDMExternalStoreImportService.java:1050)
exo_1    |      at org.exoplatform.services.organization.externalstore.IDMExternalStoreImportService.checkModifiedEntitiesOfType(IDMExternalStoreImportService.java:265)
exo_1    |      at org.exoplatform.services.organization.externalstore.IDMExternalStoreImportService.checkModifiedEntitiesOfType(IDMExternalStoreImportService.java:233)
exo_1    |      at org.exoplatform.services.organization.externalstore.IDMExternalStoreImportService.importAllModifiedEntitiesToQueue(IDMExternalStoreImportService.java:140)
exo_1    |      at org.exoplatform.services.organization.externalstore.job.IDMEntitiesImportJob.execute(IDMEntitiesImportJob.java:20)
exo_1    |      ... 2 common frames omitted
exo_1    | Caused by: org.picketlink.idm.api.query.QueryException: Failed to execute query
exo_1    |      at org.picketlink.idm.impl.api.query.UserQueryExecutorImpl.execute(UserQueryExecutorImpl.java:220)
exo_1    |      at org.picketlink.idm.impl.api.query.UserQueryExecutorImpl.list(UserQueryExecutorImpl.java:245)
exo_1    |      at org.picketlink.idm.impl.api.session.IdentitySessionImpl.list(IdentitySessionImpl.java:364)
exo_1    |      at org.exoplatform.services.organization.idm.externalstore.IDMExternalStoreListAccess.lambda$load$0(IDMExternalStoreListAccess.java:64)
exo_1    |      at org.exoplatform.services.organization.idm.externalstore.PicketLinkIDMExternalStoreService.lambda$executeOnExternalStore$9(PicketLinkIDMExternalStoreService.java:411)
exo_1    |      at org.exoplatform.services.organization.idm.externalstore.PicketLinkIDMExternalStoreService.lambda$new$0(PicketLinkIDMExternalStoreService.java:86)
exo_1    |      ... 10 common frames omitted
exo_1    | Caused by: org.picketlink.idm.common.exception.IdentityException: IdentityObject search failed.
exo_1    |      at org.picketlink.idm.impl.store.ldap.ExoLDAPIdentityStoreImpl.findIdentityObject(ExoLDAPIdentityStoreImpl.java:470)
exo_1    |      at org.gatein.portal.idm.impl.repository.ExoFallbackIdentityStoreRepository.findIdentityObject(ExoFallbackIdentityStoreRepository.java:252)
exo_1    |      at org.picketlink.idm.impl.api.session.managers.PersistenceManagerImpl.findUser(PersistenceManagerImpl.java:465)
exo_1    |      at org.picketlink.idm.impl.api.query.UserQueryExecutorImpl.execute(UserQueryExecutorImpl.java:73)
exo_1    |      ... 15 common frames omitted
exo_1    | Caused by: org.picketlink.idm.common.exception.IdentityException: Could not obtain LDAP connection:
exo_1    |      at org.picketlink.idm.impl.store.ldap.LDAPIdentityStoreImpl.getLDAPContext(LDAPIdentityStoreImpl.java:4106)
exo_1    |      at org.picketlink.idm.impl.store.ldap.LDAPIdentityStoreImpl.searchIdentityObjects(LDAPIdentityStoreImpl.java:3733)
exo_1    |      at org.picketlink.idm.impl.store.ldap.ExoLDAPIdentityStoreImpl.searchIdentityObjects(ExoLDAPIdentityStoreImpl.java:495)
exo_1    |      at org.picketlink.idm.impl.store.ldap.ExoLDAPIdentityStoreImpl.findIdentityObject(ExoLDAPIdentityStoreImpl.java:440)
exo_1    |      ... 18 common frames omitted
exo_1    | Caused by: org.picketlink.idm.common.exception.IdentityException: Could not create LdapContext
exo_1    |      at org.picketlink.idm.impl.store.ldap.LDAPIdentityStoreSessionImpl.getSessionContext(LDAPIdentityStoreSessionImpl.java:177)
exo_1    |      at org.picketlink.idm.impl.store.ldap.LDAPIdentityStoreImpl.getLDAPContext(LDAPIdentityStoreImpl.java:4097)
exo_1    |      ... 21 common frames omitted
exo_1    | Caused by: javax.naming.AuthenticationException: [LDAP: error code 49 - 80090308: LdapErr: DSID-0C090434, comment: AcceptSecurityContext error, data 52e, v4f7c]
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.mapErrorCode(LdapCtx.java:3260)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:3206)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:2992)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2906)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:348)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxFromUrl(LdapCtxFactory.java:229)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:189)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:247)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:154)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:84)
exo_1    |      at java.naming/javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:732)
exo_1    |      at java.naming/javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:305)
exo_1    |      at java.naming/javax.naming.InitialContext.init(InitialContext.java:236)
exo_1    |      at java.naming/javax.naming.ldap.InitialLdapContext.<init>(InitialLdapContext.java:154)
exo_1    |      at org.picketlink.idm.impl.store.ldap.LDAPIdentityStoreSessionImpl.getLdapContext(LDAPIdentityStoreSessionImpl.java:160)
exo_1    |      at org.picketlink.idm.impl.store.ldap.LDAPIdentityStoreSessionImpl.getSessionContext(LDAPIdentityStoreSessionImpl.java:167)
exo_1    |      ... 22 common frames omitted

**Reply by rdenarie** [reply]
You said you're able to connect in the platform.
Do you use a AD account for that  ?
or root account ?

In fact, if the configuration is ok, you should only able to use ad account to log into the platform.
In addition the synchronization can take few minutes to start, to see all users. The job runs each 10 minutes.
In addition, can you add this empty property in your properties file :
exo.ldap.groups.base.dn= 

When you login with a AD user, the synchronization for this user is forced, and we should see log in the log file.

**Reply by rdenarie** [reply]
LDAP : error code 49 means authentication problem.

In the admin DN, you should escape the \ : 
exo.ldap.admin.dn=EXO\\Administrateur

**Reply by xBlueDragonn** [reply]
yes i log in with root account to check if user are up.

i change how to declare admin login from ROOT\Administrateur to Administrateur@ROOT and i have now this error:
Caused by: javax.naming.NameNotFoundException: [LDAP: error code 32 - 0000208D: NameErr: DSID-0310028B, problem 2001 (NO_OBJECT), data 0, best match of:
exo_1    |      'DC=exo,DC=local'
exo_1    | ]
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.mapErrorCode(LdapCtx.java:3285)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:3206)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:2997)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.searchAux(LdapCtx.java:1875)
exo_1    |      at java.naming/com.sun.jndi.ldap.LdapCtx.c_search(LdapCtx.java:1798)
exo_1    |      at java.naming/com.sun.jndi.toolkit.ctx.ComponentDirContext.p_search(ComponentDirContext.java:392)
exo_1    |      at java.naming/com.sun.jndi.toolkit.ctx.PartialCompositeDirContext.search(PartialCompositeDirContext.java:358)
exo_1    |      at java.naming/javax.naming.directory.InitialDirContext.search(InitialDirContext.java:305)
exo_1    |      at org.picketlink.idm.impl.store.ldap.LDAPIdentityStoreImpl.searchLDAP(LDAPIdentityStoreImpl.java:3951)
exo_1    |      at org.picketlink.idm.impl.store.ldap.LDAPIdentityStoreImpl.searchIdentityObjects(LDAPIdentityStoreImpl.java:3775)
exo_1    |      at org.picketlink.idm.impl.store.ldap.ExoLDAPIdentityStoreImpl.searchIdentityObjects(ExoLDAPIdentityStoreImpl.java:495)
exo_1    |      at org.picketlink.idm.impl.store.ldap.ExoLDAPIdentityStoreImpl.findIdentityObject(ExoLDAPIdentityStoreImpl.java:440)
exo_1    |      ... 18 common frames omitted

It seem to mean that what i'm looking for do not exist

**Reply by rdenarie** [reply]
What is your AD Domain name ? 
What is the technical user with which eXo will request AD ? 

let say that the domain is EXO, and the user is Administrateur, the property should be 
exo.ldap.admin.dn=EXO\\Administrateur
This user is the user with which exoplatform with make synchronization requests. It is not the user with which you will login exo.

**Reply by xBlueDragonn** [reply]
yes my domain is exo.local and Administrateur is the administrator of my Active directory server

**Reply by xBlueDragonn** [reply]
Thanks for your help it finally work 
So th problem was first with ROOT\Administrateur i got error 49 so i switch with Administrateur@exo.local that resolved the error
and for the error 32, just need to change exo.ldap.users.base.dn=ou=users1,dc=company,dc=com to exo.ldap.users.base.dn=**cn**=users1,dc=company,dc=com

and it finnaly works 
Thanks again for your help [rdenarie](https://github.com/rdenarie)

**Comment by rdenarie** [comment]
Ok, I mark it as resolved !
Thanks

**Comment by aurelienfasquelle-lang** [comment]
Hello everyone,
I successfully connected eXo Platform to my Active Directory, but the mapping of additional user information is not working. Does anyone know why or how to fix it?

**Reply by rdenarie** [reply]
Hello @aurelienfasquelle-lang 

Can you post your question on community.exoplatform.com ?
You will have more chance to get an anwser 

Thanks

**Reply by aurelienfasquelle-lang** [reply]
Merci beaucoup pour votre réponse
